dali/kernels/kernel_manager.h

```diff
@@ -0,0 +1,237 @@
+    assert(static_cast<size_t>(thread_idx) < scratchpads.size());
+    auto &sa = GetScratchpadAllocator(thread_idx);
+    Run<Kernel>(sa, instance_idx, context, std::forward<OutInArgs>(out_in_args)...);
+  }
+
+  /// @brief Calls Run on specified kernel instance using Scratchpad for given thread.
+  ///
+  /// @param sa             - scratchpad allocator; memory will be reserved in it to satisfy
+  ///                         instance's requirements
+  /// @param instance_idx   - kernel instance index; typically corresponds
+  ///                         to sample index (for per-sample kernels) or minibatch index
+  /// @param context        - context for the kernel
+  ///                         * should contain valid CUDA stream for GPU kernels;
+  ///                         * scratchpad pointer is overriden with a scratchpad
+  ///                           created from `sa`
+  /// @param out_in_args    - pack of arguments (outputs, inputs, arguments) used in Kernel::Run
+  template <typename Kernel, typename... OutInArgs>
+  void Run(ScratchpadAllocator &sa,
+           int instance_idx,
+           KernelContext &context,
+           OutInArgs &&...out_in_args) {
+    assert(instance_idx >= 0 &&
+           static_cast<size_t>(instance_idx) < NumInstances() &&
+           "Kernel instance index (instance_idx) out of range");
+    auto &inst = instances[instance_idx];
+    ReserveScratchpad(sa, inst.requirements.scratch_sizes);
+    auto scratchpad = sa.GetScratchpad();
+    auto *old_scratchpad = context.scratchpad;
+    context.scratchpad = &scratchpad;
+    inst.get<Kernel>().Run(context, std::forward<OutInArgs>(out_in_args)...);
+    context.scratchpad = old_scratchpad;
+  }
+
+  /// @brief Makes sure ScratchpadAllocator can accommodate `sizes`
+  ///
+  /// @param sa     - scratchpad allocator to reserve
+  /// @param sizes  - requested minimum size
+  ///
+  /// The manager maintains a lifetime maximum of sizes requested.
+  /// If reallocation is necessary, it allocates `sizes` or that maximum
+  /// whichever is larger.
+  void ReserveScratchpad(ScratchpadAllocator &sa, const ScratchSizes &sizes);
+
+  /// @brief Calls ReserveScratchpad on ScratchpadAllocator associated with given thread_idx
+  inline void ReserveScratchpad(int thread_idx, const ScratchSizes &sizes) {
+    ReserveScratchpad(GetScratchpadAllocator(thread_idx), sizes);
+  }
+
+  /// @brief Sets a memory size hint for allocating scratchpad memory
+  ///
+  /// All calls to ScratchpadAllocator::Reserve followint this call will request at least
+  /// bytes memory for given allocation type.
+  void SetMemoryHint(AllocType type, size_t bytes) {
+    int alloc_idx = static_cast<int>(type);
+    if (bytes > max_scratch_sizes[alloc_idx])
+      max_scratch_sizes[alloc_idx] = bytes;
+  }
+
+ private:
+  SmallVector<AnyKernelInstance, 1> instances;
+  SmallVector<ScratchpadAllocator, 1> scratchpads;
```

Comment:
Why `1`?

Reply:
These are not very lightweight objects - however, for the case of GPU operators, this will most likely be 1 instance (whole batch) and 1 thread.

Reply:
Makes sense.